src/common/state/task_table.h

```
@@ -1,20 +1,130 @@
-/* Callback for getting an entry from the task table. Task spec will be freed
- * by the system after the callback */
-typedef void (*task_table_callback)(task_spec *task, void *context);
+/* Callback called when a task table write operation completes. */
+typedef void (*task_table_done_callback)(task_id task_id, void *user_context);
 
-/* Get specific task from the task table. */
-status task_table_get_task(db_handle *db,
-                           task_id task_id,
-                           task_table_callback callback,
-                           void *context);
+/* Callback called when a task table read operation completes. */
+typedef void (*task_table_get_callback)(task *task, void *user_context);
 
-#endif /* TASK_TABLE_H */
+/**
+ * Get a task's entry from the task table.
+ *
+ * @param db_handle Database handle.
+ * @param task_id The ID of the task we want to look up.
+ * @param retry Information about retrying the request to the database.
+ * @param done_callback Function to be called when database returns result.
+ * @param user_context Data that will be passed to done_callback and
+ *        fail_callback.
+ * @return Void.
+ */
+void task_table_get_task(db_handle *db,
+                         task_id task_id,
+                         retry_info *retry,
+                         task_table_get_callback done_callback,
+                         void *user_context);
+
+/**
+ * Add a task entry, including task spec and scheduling information, to the
+ * task table. This will overwrite any task already in the task table with the
+ * same
+ * task ID.
+ *
+ * @param db_handle Database handle.
+ * @param task The task entry to add to the table.
+ * @param retry Information about retrying the request to the database.
+ * @param done_callback Function to be called when database returns result.
+ * @param user_context Data that will be passed to done_callback and
+ *        fail_callback.
+ * @return Void.
+ */
+void task_table_add_task(db_handle *db_handle,
```

Right, but the documentation says "this will override anything already in the table", so I'm wondering if there are situations where we expect that to happen. If not, then it's a bug so we can throw a fatal error, right?